### PR TITLE
Randomness Stream typing

### PR DESCRIPTION
**Pull request by patricktnast**
### Description

- *Category*: 
feature
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5398

### Changes and notes
 
use float | object for the probability sequence when possible, otherwise (for pd.Series) use Any.

### CODE REVIEW

File: src/vivarium/framework/randomness/stream.py [hunk 1]
```diff
@@ -25,19 +24,26 @@
 
 """
 
+from __future__ import annotations
+
 import hashlib
-from typing import Any, Callable, List, Optional, Tuple, Union
+from typing import Any, Callable, TypeVar
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:30]: Nit: import `Callable` as `from collections.abc import Callable`

File: src/vivarium/framework/randomness/stream.py [hunk 2]
```diff
@@ -254,18 +260,28 @@ def filter_for_probability(
         if population.empty:
             return population
 
-        index = population if isinstance(population, pd.Index) else population.index
+        if isinstance(population, pd.Index):
+            index = population
+        else:
+            index = population.index
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:266]: Why was this change necessary?

File: src/vivarium/framework/randomness/stream.py [hunk 3]
```diff
@@ -25,19 +24,26 @@
...
 
 from vivarium.framework.randomness.exceptions import RandomnessError
 from vivarium.framework.randomness.index_map import IndexMap
 from vivarium.framework.utilities import rate_to_probability
+from vivarium.types import ClockTime, NumericArray
 
 RESIDUAL_CHOICE = object()
 
+# TODO: Parameterizing pandas objects fails below python 3.12
+P = TypeVar("P", pd.DataFrame, pd.Series, pd.Index)  # type: ignore [type-arg]
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:45]: Can you give this TypeVar a more informative name? Maybe something like `PandasObject`?

File: src/vivarium/framework/randomness/stream.py [hunk 4]
```diff
@@ -305,12 +321,12 @@ def choice(
 
     def sample_from_distribution(
         self,
-        index: pd.Index,
-        distribution: Optional[stats.rv_continuous] = None,
-        ppf: Optional[Callable[[pd.Series, dict[str, Any]], pd.Series]] = None,
+        index: pd.Index[int],
+        distribution: stats.rv_continuous | None = None,
+        ppf: Callable[[pd.Series[Any], dict[str, Any]], pd.Series[Any]] | None = None,
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:326]: I'm pretty sure a `ppf` needs to have floats as input. So you should be able to type this as 
```suggestion
        ppf: Callable[[pd.Series[float], dict[str, Any]], pd.Series[Any]] | None = None,
```

File: src/vivarium/framework/randomness/stream.py [hunk 5]
```diff
@@ -254,18 +260,28 @@ def filter_for_probability(
...
-        p: Optional[Union[List, Tuple, np.ndarray, pd.Series]] = None,
-        additional_key: Optional[Any] = None,
-    ) -> pd.Series:
+        index: pd.Index[int],
+        choices: list[Any] | tuple[Any] | npt.NDArray[Any] | pd.Series[Any],
+        p: (
+            list[float | object]
+            | tuple[float | object]
+            | npt.NDArray[np.number[npt.NBitBase] | np.object_]
+            | pd.Series[Any]
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:280]: This should be `pd.Series[float]`?

File: src/vivarium/framework/randomness/stream.py [hunk 6]
```diff
@@ -351,10 +367,16 @@ def __repr__(self) -> str:
...
-    choices: Union[List, Tuple, np.ndarray, pd.Series],
-    p: Optional[Union[List, Tuple, np.ndarray, pd.Series]] = None,
-) -> pd.Series:
+    draws: pd.Series[float],
+    choices: list[Any] | tuple[Any] | npt.NDArray[Any] | pd.Series[Any],
+    p: (
+        list[float | object]
+        | tuple[float | object]
+        | npt.NDArray[np.number[npt.NBitBase] | np.object_]
+        | pd.Series[Any]
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:376]: float?

File: src/vivarium/framework/randomness/stream.py [hunk 7]
```diff
@@ -389,24 +411,31 @@ def _choice(
...
 
 def _normalize_shape(
-    p: Union[List, Tuple, np.ndarray, pd.Series],
-    index: Union[pd.Index, pd.MultiIndex],
-) -> np.ndarray:
+    p: (
+        list[float | object]
+        | tuple[float | object]
+        | npt.NDArray[np.number[npt.NBitBase] | np.object_]
+        | pd.Series[Any]
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:435]: float?

File: src/vivarium/framework/randomness/stream.py [hunk 8]
```diff
@@ -389,24 +411,31 @@ def _choice(
...
+    ),
+    index: pd.Index[int] | pd.MultiIndex,
+) -> npt.NDArray[Any]:
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/stream.py:438]: shouldn't this be `npt.NDArray[np.number[npt.NBitBase] | np.object_]`?

File: src/vivarium/framework/randomness/stream.py (hunk 2 above)
Comment (patricktnast) [reply src/vivarium/framework/randomness/stream.py:266]: mypy gave "pd.Index has no attribute index" even though you think it would be smart enough.

File: src/vivarium/framework/randomness/stream.py (hunk 8 above)
Comment (patricktnast) [reply src/vivarium/framework/randomness/stream.py:438]: yes

**Review by rmudambi — APPROVED** [review]

**Review by albrja — APPROVED** [review]
